RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco2D.cc

```diff
@@ -0,0 +1,654 @@
+               indexxy[0][tpixel] = j;
+               indexxy[1][tpixel] = i-1;
+               clusxy[j][i-1] = pseudopix;
+               pixel[tpixel] = pseudopix;
+               sigma2[tpixel] = pseudopix2;
+               ++tpixel;
+            }
+            if((j+1) != jmisslow) {
+               if(clusxy[j+1][i-1] < pseudopix) {
+                  indexxy[0][tpixel] = j+1;
+                  indexxy[1][tpixel] = i-1;
+                  clusxy[j+1][i-1] = pseudopix;
+                  pixel[tpixel] = pseudopix;
+                  sigma2[tpixel] = pseudopix2;
+                  ++tpixel;
+               }
+            }
+         }
+      }
+      
+      
+      //   		 printf("** 2D-cluster, cotalpha = %f, cotbeta = %f **\n", cotalpha, cotbeta);
+      
+      //		 for (k=1; k < BXM3; ++k) {
+      //			 printf("%5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f %5.0f \n",
+      //clusxy[k][1],clusxy[k][2],clusxy[k][3],clusxy[k][4],clusxy[k][5],clusxy[k][6],clusxy[k][7],clusxy[k][8],clusxy[k][9],
+      //clusxy[k][10],clusxy[k][11],clusxy[k][12],clusxy[k][13],clusxy[k][14],clusxy[k][15],clusxy[k][16],clusxy[k][17],clusxy[k][18],
+      //					  clusxy[k][19],clusxy[k][20],clusxy[k][21]);
+      //		 }
+      
+      
+      
+      // Calculate chi2 over a grid of several seeds and choose the smallest
+      
+      chi2min[ipass] = 1000000.f;
+      float chi2, qtemplate, qactive, qtfrac = 0.f, x2D = 0.f, y2D = 0.f;
+      for(int is = -4; is<5; ++is) {
+         for(int js = -2; js<3; ++js) {
+            float xtry = x0 + js*xsize/4.;
+            float ytry = y0 + is*ysize/4.;
+//      for(int is = -8; is<9; ++is) {
+//         for(int js = -4; js<5; ++js) {
+//            float xtry = x0 + js*xsize/8.;
+//            float ytry = y0 + is*ysize/8.;
+            chi2 = 0.f;
+            qactive = 0.f;
+            for(j=0; j<BXM2; ++j) {for(i=0; i<BYM2; ++i) {template2d[j][i] = 0.f;}}
+            templ2D.xytemp(xtry, ytry, ydouble, xdouble, template2d, false, dpdx2d, qtemplate);
+            for(k=0; k<tpixel; ++k) {
+               int jpix = indexxy[0][k];
+               int ipix = indexxy[1][k];
+               float qtpixel = template2d[jpix][ipix];
+               float pt = pixel[k]-qtpixel;
+               chi2 += pt*pt/sigma2[k];
+               if(k < npixel) {qactive += qtpixel;}
+            }
+            if(chi2 < chi2min[ipass]) {
+               chi2min[ipass] = chi2;
+               x2D = xtry;
+               y2D = ytry;
+               qtfrac = qactive/qtemplate;
+            }
+         }
+      }
+      //      printf("jsel/isel = %d/%d\n", jsel, isel);
+      
+      
+      //   printf("\n begin iterative minimization \n");
+
+      
+      int niter = 0;
+      float xstep = 1.0f, ystep = 1.0f;
+      float minv11 = 1000.f, minv12 = 1000.f, minv22 = 1000.f;
+      //      chi2min[ipass] = 100000.f;

```

Comment:
Are these commented out lines needed? If so, please add another line of comment which explain why; otherwise just remove them

Reply:
I don't know.  I'll need to discuss with Morris.  I propose to postpone this until later.